### ISSUE TITLE
toplevel_ref_arg triggers on macro-generated code

### ISSUE THREAD

**Issue by jplatte** [rust-lang/rust-clippy#5849]
Not sure whether this is intended behaviour, but currently pretty much all SQLx `query_as!` invocations trigger this lint, e.g.

```rust
sqlx::query_as!(SomeType, "SELECT * FROM table WHERE id = $1", id)
    .fetch_one(pool)
    .await
```

> `ref` on an entire `let` pattern is discouraged, take a reference with `&` instead

### Meta

- `cargo clippy -V`: `clippy 0.0.212 (6c8927b 2020-07-26)`
- `rustc -Vv`:
  ```
  rustc 1.47.0-nightly (6c8927b0c 2020-07-26)
  binary: rustc
  commit-hash: 6c8927b0cf80ceee19386026cf9d7fd4fd9d486f
  commit-date: 2020-07-26
  host: x86_64-unknown-linux-gnu
  release: 1.47.0-nightly
  LLVM version: 10.0
  ```

**Comment by abonander** [comment]
This can be trivially fixed in SQLx, but in general it would be nice to know how we could avoid triggering lint warnings in generated code which the user has no control over.

**Comment by abonander** [comment]
It's also worth mentioning that this warning was produced with `sqlx 0.4.0-beta.1`

**Comment by Ryan1729** [comment]
> it would be nice to know how we could avoid triggering lint warnings in generated code which the user has no control over.

Depending on what form the macro output takes, you could add an `#[allow(clippy::foo)]` annotation to it.

For instance, if the output is an expression, you can wrap the output in [a block and add an inner attribute at the top](https://doc.rust-lang.org/reference/expressions/block-expr.html).

[Playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=5d653c3b6a23ab8650bdb3102cc78f06)

**Comment by ebroto** [comment]
To fix this issue, we can use the `utils::in_macro` function to check if we have to bail out. This lint is found in the `misc.rs` module. 

For the tests, we should create two macros, and use them:
* one that generates a function using `ref` in parameter position. Example:
* another that generates a `let` binding using ref.

For example:
```rust
macro_rules! gen_binding {
    () => {
        let ref _y = 42;
    }
}

macro_rules! gen_function {
    () => {
        fn fun_example(ref _x: usize) {}
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8a47893d982e60e390bf5044d4625709))
